aidies-project/client/src/components/LoginPanel.js

```diff
@@ -1,45 +1,117 @@
 import React, { useState } from 'react';
-import { Redirect, NavLink } from 'react-router-dom';
+import { Redirect, NavLink, useHistory } from 'react-router-dom';
 import { useDispatch, useSelector } from 'react-redux';
 import { login } from '../store/authentication';
-// import { Redirect } from 'react-router-dom';
+
+import {Button, 
+        CssBaseline, 
+        TextField, 
+        Link, 
+        Grid, 
+        Typography, 
+        Container} from '@material-ui/core';
+
+import { makeStyles } from '@material-ui/core/styles';
+
+const useStyles = makeStyles((theme) => ({
+    paper: {
+        marginTop: theme.spacing(8),
+        display: 'flex',
+        flexDirection: 'column',
+        alignItems: 'center',
+    },
+    avatar: {
+        margin: theme.spacing(1),
+        backgroundColor: theme.palette.secondary.main,
+    },
+    form: {
+        width: '100%', // Fix IE 11 issue.
+        marginTop: theme.spacing(1),
+    },
+    submit: {
+        margin: theme.spacing(3, 0, 2),
+        backgroundColor: '#222',
+        color: '#fff',
+    },
+    link: {
+        color: '#222',
+    }
+}));
 
 const LoginPanel = (props) => {
     const [email, setEmail] = useState('demo@example.com');
     const [password, setPassword] = useState('password');
     const dispatch = useDispatch();
-    const user = useSelector(state => state.authentication.user);
-    
+    const user = useSelector(state => state.authentication.user.id);
+    const location = useSelector(state => state.campaign.previousLocation);
+    const history = useHistory();
+    const classes = useStyles();
+
     const handleSubmit = (e) =>{
         e.preventDefault();
         dispatch(login(email, password));
     }
-    const updateEmail = (e)=>{
-        setEmail(e.target.value);
-    }
-    
-    const updatePassword = (e)=>{
-        setPassword(e.target.value);
-    }
-    
-    if (user) {
-        return <Redirect to="/" />
+    if (user!== null) {
+        if(!location){
+            return <Redirect to="/" />
+        }
+        return <Redirect to={location} />
```

Comment:
Congratulations on setting up this redirection! Now that you have it working, I would move the `user` and `location` `useSelector` hooks into a `<ProtectedRoute>` higher order component. 

The `<ProtectedRoute>` would take care of
- using the `user` and `location` from state
- redirecting to `/` if there is no `location` and the user is not null
- redirection to `location` if the user is not null
- or rendering the `Component` prop

Reply:
Would that be in the `App` component? or within the same component?